index.js

```diff
@@ -585,20 +627,73 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<Array<Keyring>>} The keyrings.
    */
-  async unlockKeyrings(password) {
+  async unlockKeyrings(password, encryptedKey) {
     const encryptedVault = this.store.getState().vault;
     if (!encryptedVault) {
       throw new Error('Cannot unlock without a previous vault.');
     }
 
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    // MV3: If the separator string is in the vault string, the user has already migrated
+    // from the previous password-only model
+    let vault = null;
+    if (encryptedVault && encryptedVault.includes(VAULT_SEPARATOR)) {
+      const [, salt] = encryptedVault.split(VAULT_SEPARATOR);
+
+      if (password) {
+        this.encryptedKey = this._generateEncryptedKey(password, salt);
+      } else if (encryptedKey) {
+        this.encryptedKey = encryptedKey;
+      } else {
+        throw new Error(
+          'No way to decrypt a salted vault without a password or encrypted key',
+        );
+      }
+
+      vault = await this.encryptor.decrypt(this.encryptedKey, encryptedVault);
+    } else {
+      vault = await this.encryptor.decrypt(password, encryptedVault);
+    }
+
     await Promise.all(vault.map(this._restoreKeyring.bind(this)));
+
     await this._updateMemStoreKeyrings();
+
+    // MV3: If we're provided a password, we should persist keyrings
+    // so that we can either (1) migrate or (2) create a new salt
+    if (password) {
+      await this.persistAllKeyrings(password);
+    }
+
     return this.keyrings;
   }
 
+  // MV3:  Generates the encrypted key
+  _generateEncryptedKey(password, salt) {
+    // TODO: How complicaated do we want this to be?
+    return window.btoa(password + salt);
+  }
+
+  // MV3:  Returns the encrypted key so it's accessible from the extension
+  getEncryptedKey() {
+    return this.encryptedKey;
+  }
+
+  // MV3:  Creates a salt to be used in encrypted key unlocking
+  // https://github.com/MetaMask/browser-passworder/blob/d24b66934ab4d39d97ccff9ff0c3d25f86f1a141/src/index.ts#L180
+  _generateSalt(byteCount = 32) {
+    const view = new Uint8Array(byteCount);
+    global.crypto.getRandomValues(view);
+    // Uint8Array is a fixed length array and thus does not have methods like pop, etc
+    // so TypeScript complains about casting it to an array. Array.from() works here for
+    // getting the proper type, but it results in a functional difference. In order to
+    // cast, you have to first cast view to unknown then cast the unknown value to number[]
+    // TypeScript ftw: double opt in to write potentially type-mismatched code.
+    const b64encoded = window.btoa(String.fromCharCode.apply(null, view));
```

Comment:
We could use `Buffer.from(...).toString('base64')` here instead of `window.btoa`. That would ensure this library still works in a Node.js context.

Reply:
Oh, i see this was taken from `browser-passworder`, which is only meant to be used by browsers.

Any reason this was inlined, rather than calling `this.encryptor.generateSalt`?